'SPI: Implementing tenant control 0.1-SNAPSHOT' does not compile

Hello

2 days ago in 2 machines (MacOS and Windows) I did do an update about the source code through Git.
When I execute: `mvn clean compile` in the root project, throws:
```
[INFO] SPI 0.1-SNAPSHOT ................................... SUCCESS [  0.003 s]
[INFO] Discovery SPI 0.1-SNAPSHOT ......................... SUCCESS [  0.110 s]
[INFO] SPI: Split-brain healing 0.1-SNAPSHOT .............. SUCCESS [  0.124 s]
[INFO] SPI: Implementing tenant control 0.1-SNAPSHOT ...... FAILURE [  0.115 s]
[INFO] Transactions 0.1-SNAPSHOT .......................... SKIPPED
[INFO] Transactions - Transaction Basics 0.1-SNAPSHOT ..... SKIPPED
[INFO] Transactions - Transactional Task 0.1-SNAPSHOT ..... SKIPPED
[INFO] Transactions - XA Transaction 0.1-SNAPSHOT ......... SKIPPED
[INFO] Variable replacers 0.1-SNAPSHOT .................... SKIPPED
[INFO] genericRecord 0.1-SNAPSHOT ......................... SKIPPED
[INFO] generic-record-cluster 0.1-SNAPSHOT ................ SKIPPED
[INFO] generic-recod-client 0.1-SNAPSHOT .................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:02 min
[INFO] Finished at: 2020-12-16T10:01:13-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project tenant-control: Compilation failure: Compilation failure: 
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java:[9,8] multitenant.MultiTenantControl is not abstract and does not override abstract method clearThreadContext() in com.hazelcast.spi.tenantcontrol.TenantControl
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java:[17,5] method does not override or implement a method from a supertype
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java:[25,16] incompatible types: multitenant.TenantCloseable cannot be converted to com.hazelcast.spi.tenantcontrol.TenantControl.Closeable
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java:[28,5] method does not override or implement a method from a supertype
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java:[7,8] multitenant.MultiTenantControlFactory is not abstract and does not override abstract method isClassesAlwaysAvailable() in com.hazelcast.spi.tenantcontrol.TenantControlFactory
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java:[9,5] method does not override or implement a method from a supertype
[ERROR] /Users/.../hazelcast-code-samples/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java:[11,33] cannot find symbol
[ERROR]   symbol:   method getDistributedObjectName()
[ERROR]   location: variable event of type com.hazelcast.spi.tenantcontrol.DestroyEventContext
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :tenant-control

```

**Note**: in these 2 machines the repositories were deleted and used `project -> Maven -> Update Project` (it works fine) and then `mvn clean compile` (here fails)

--- a/spi/tenant-control/src/main/java/multitenant/Bootstrap.java
+++ b/spi/tenant-control/src/main/java/multitenant/Bootstrap.java
@@ -15,6 +15,7 @@ import static java.util.Collections.singletonList;
 public class Bootstrap {
     public static void main(String[] args) {
         ClassLoader parent = Bootstrap.class.getClassLoader();
+        MultiTenantControlFactory.TENANT_CACHE.set("apps.app1");
         MultiTenantClassLoader multitenantClassLoader = new MultiTenantClassLoader(parent);
         // app1 classloader excludes app2, defines apps.app1.* classes on its own
         multitenantClassLoader.addClassLoader("apps.app1", new TenantClassloader(

--- a/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
+++ b/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
@@ -15,7 +15,7 @@ public class MultiTenantControl implements TenantControl {
     private String cacheName;
     private DestroyEventContext destroyEventContext;
 
-    public MultiTenantControl(String cacheName) {
+    MultiTenantControl(String cacheName) {
         this.cacheName = cacheName;
     }
 
@@ -41,7 +41,7 @@ public class MultiTenantControl implements TenantControl {
     @Override
     public void unregisterObject() {
         destroyEventContext.tenantUnavailable();
-        destroyEventContext = () -> {};
+        destroyEventContext = () -> { };
     }
 
     @Override

--- a/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
+++ b/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
@@ -15,6 +15,10 @@ public class MultiTenantControl implements TenantControl {
     private String cacheName;
     private DestroyEventContext destroyEventContext;
 
+    public MultiTenantControl(String cacheName) {
+        this.cacheName = cacheName;
+    }
+
     @Override
     public Closeable setTenant() {
         ClassLoader original;
@@ -23,6 +27,7 @@ public class MultiTenantControl implements TenantControl {
             original = Thread.currentThread().getContextClassLoader();
             Thread.currentThread().setContextClassLoader(CLASS_LOADER_PER_PREFIX.get(cacheName));
         } else {
+            System.out.println("No Tenant in the thread context");
             original = null;
         }
         return new TenantCloseable(original)::close;

--- a/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
+++ b/spi/tenant-control/src/main/java/multitenant/MultiTenantControl.java
@@ -1,32 +1,60 @@
 package multitenant;
 
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.spi.tenantcontrol.DestroyEventContext;
 import com.hazelcast.spi.tenantcontrol.TenantControl;
+import com.hazelcast.spi.tenantcontrol.Tenantable;
+import java.io.IOException;
 
-import java.io.Closeable;
 
 import static multitenant.MultiTenantClassLoader.CLASS_LOADER_PER_PREFIX;
 
 public class MultiTenantControl implements TenantControl {
 
-    private final String cacheName;
-
-    MultiTenantControl(String cacheName) {
-        this.cacheName = cacheName;
-    }
+    private String cacheName;
+    private DestroyEventContext destroyEventContext;
 
     @Override
-    public Closeable setTenant(boolean createRequestScope) {
-        ClassLoader original = null;
+    public Closeable setTenant() {
+        ClassLoader original;
         if (cacheName != null) {
             System.out.println("Setting class loader for cache " + cacheName + " / " + CLASS_LOADER_PER_PREFIX.get(cacheName));
             original = Thread.currentThread().getContextClassLoader();
             Thread.currentThread().setContextClassLoader(CLASS_LOADER_PER_PREFIX.get(cacheName));
+        } else {
+            original = null;
         }
-        return new TenantCloseable(original);
+        return new TenantCloseable(original)::close;
     }
 
     @Override
-    public void unregister() {
-        // nothing to do
+    public void registerObject(DestroyEventContext destroyEventContext) {
+        this.destroyEventContext = destroyEventContext;
+    }
+
+    @Override
+    public void unregisterObject() {
+        destroyEventContext.tenantUnavailable();
+        destroyEventContext = () -> {};
+    }
+
+    @Override
+    public boolean isAvailable(Tenantable tenantable) {
+        return true;
+    }
+
+    @Override
+    public void clearThreadContext() {
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeUTF(cacheName);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        cacheName = in.readUTF();
     }
 }

--- a/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java
+++ b/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java
@@ -2,12 +2,14 @@ package multitenant;
 
 import com.hazelcast.spi.tenantcontrol.TenantControl;
 import com.hazelcast.spi.tenantcontrol.TenantControlFactory;
+import java.util.concurrent.atomic.AtomicReference;
 
 public class MultiTenantControlFactory implements TenantControlFactory {
+    static final AtomicReference<String> TENANT_CACHE = new AtomicReference<>();
 
     @Override
     public TenantControl saveCurrentTenant() {
-        return new MultiTenantControl();
+        return new MultiTenantControl(TENANT_CACHE.getAndSet(null));
     }
 
     @Override

--- a/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java
+++ b/spi/tenant-control/src/main/java/multitenant/MultiTenantControlFactory.java
@@ -1,14 +1,17 @@
 package multitenant;
 
-import com.hazelcast.spi.tenantcontrol.DestroyEventContext;
 import com.hazelcast.spi.tenantcontrol.TenantControl;
 import com.hazelcast.spi.tenantcontrol.TenantControlFactory;
 
 public class MultiTenantControlFactory implements TenantControlFactory {
 
     @Override
-    public TenantControl saveCurrentTenant(DestroyEventContext event) {
-        String cacheName = event.getDistributedObjectName();
-        return new MultiTenantControl(cacheName);
+    public TenantControl saveCurrentTenant() {
+        return new MultiTenantControl();
+    }
+
+    @Override
+    public boolean isClassesAlwaysAvailable() {
+        return false;
     }
 }

